### PR TITLE
MapWriter benchmark

### PR DESCRIPTION
**Pull request by laithsakka**
Summary:
**TLDR:**
- The new map writer interface is 6X-8X faster than the old one.
- Simple functions performance is comparable and very close to vector implementation
that does resize per row.
- The resize + subscript is not that much faster than general and emplace, unlike
in the array writer. I will look into this if i could not make it better, I will just remove those
functions from the interface.

**Details:**
The benchmarks runs a functions with that constructs a map with no other
work to focus at map construction time. Having additional work will reduce the gap
between vector and simple.

Two vector implementations are added:
- `vector_one_resize `: does one resize for all the rows. This is not possible to do in the simple
function right now, and it's not  always possible to do even in vector function.

- `vector_resize_per_row`: resize keys and value for every processed rows.

The following simple function implementations are added:
- `simple_old` : old writer.
- `simple_general`: the general interface ( addNull and addItem).
- `simple_emplace`
- `simple_resize_subscript`

**Results:**

with nulls:
```
-bm_min_usec 1000000
============================================================================
vector_one_resize                                            1.98ms   505.57
vector_resize_per_row                                        2.49ms   402.03
simple_resize_subscript                                      2.28ms   438.92
simple_emplace                                               2.38ms   420.98
simple_general                                               2.41ms   414.58
simple_old                                                  12.78ms    78.25
============================================================================
```

no nulls:
```
 -bm_min_usec 1000000
============================================================================
vector_one_resize                                            1.08ms   925.35
vector_resize_per_row                                        1.67ms   598.42
simple_resize_subscript                                      1.56ms   640.77
simple_emplace                                               1.47ms   680.28
simple_general                                               1.46ms   685.25
simple_old                                                  11.70ms    85.44

============================================================================

 -bm_min_usec 100000000 (10^8).

============================================================================

vector_one_resize                                            1.11ms   901.87
vector_resize_per_row                                        1.56ms   639.08
simple_resize_subscript                                      1.69ms   590.73
simple_emplace                                               1.66ms   602.88
simple_general                                               1.65ms   607.12
simple_old                                                  11.30ms    88.47
============================================================================
```

**Next steps:**
- try to cache data* when key is primitive and see if that gives speedups.
- benchmark nested .i.e: Array<Map> or nested map. I am positive numbers
will be closer for those since there would be more work in general.

Differential Revision: D34289772

### CODE REVIEW

File: velox/expression/VectorUdfTypeSystem.h [hunk 1]
```diff
@@ -602,12 +602,12 @@ struct VectorWriter<ArrayProxyT<V>> {
...
 
   void reset() {
-    proxy_.valuesOffset_ = 0;
+    elementWriter_.valuesOffset_ = 0;
   }
 
   vector_t* arrayVector_ = nullptr;
 
-  exec_out_t proxy_;
+  exec_out_t elementWriter_;
```
Comment (kevinwilfong) [inline velox/expression/VectorUdfTypeSystem.h:610]: would arrayWriter_ be a better name for this?

elementWriter_ sounds like it's writing the elements of the Array, which is also what childWriter_ sounds like.

since elementWriter_ is actually the ArrayProxy (which gets renamed to ArrayWriter in a downstream change) arrayWriter_ might be more clear

File: velox/functions/prestosql/benchmarks/MapWriterBenchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
+/*
```
Comment (kevinwilfong) [inline velox/functions/prestosql/benchmarks/MapWriterBenchmark.cpp:1]: Shouldn't this have entries in CMakeLists.txt as well?

File: velox/functions/prestosql/benchmarks/MapWriterBenchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+#include <folly/init/Init.h>
+
+#include "velox/functions/Registerer.h"
+#include "velox/functions/lib/benchmarks/FunctionBenchmarkBase.h"
+
+// This macro should be set, defined in command line.
+// #define WITH_NULLS false
+
+// This file includes benchmarks that evaluate performance of map writer.
+// It benchmark a function that constructs map<int, int> (i-> i*2).
```
Comment (kevinwilfong) [inline velox/functions/prestosql/benchmarks/MapWriterBenchmark.cpp:27]: nit: benchmarks (with an s)

**Review by kevinwilfong — APPROVED** [review]

File: velox/expression/VectorUdfTypeSystem.h (hunk 1 above)
Comment (laithsakka) [reply velox/expression/VectorUdfTypeSystem.h:610]: I did this change in the a following diff that focus on replacing proxy_ naming with elementWriter_
